File: sdk/c/examples/doctor.c
/* matrix-doctor: diagnose the Matrix operator environment.
 * Usage: matrix-doctor [--binary <path>]. Prints JSON (no secrets). */
#define _POSIX_C_SOURCE 200809L

#include "mx_component.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main(int argc, char **argv) {
    const char *binary = NULL;
    for (int i = 1; i + 1 < argc; i++) {
        if (!strcmp(argv[i], "--binary"))
            binary = argv[++i];
    }
    char *rep = mx_op_doctor(binary);
    if (!rep) {
        fprintf(stderr, "doctor failed\n");
        return 1;
    }
    printf("%s\n", rep);
    int ok = strstr(rep, "\"cli_shape_ok\":true") != NULL;
    free(rep);
    return ok ? 0 : 1;
}

File: sdk/elixir/lib/matrix/doctor.ex
defmodule Matrix.Doctor do
  @moduledoc """
  `mix run -e 'Matrix.Doctor.main()'` entrypoint (see `doctor.exs`):
  prints environment diagnosis as JSON (no secrets).
  """

  def main(argv \\ System.argv()) do
    binary = parse(argv)
    rep = Matrix.Operator.doctor(binary)
    IO.puts(:erlang.iolist_to_binary(:json.encode(rep)))
    if rep["cli_shape_ok"], do: System.halt(0), else: System.halt(1)
  end

  defp parse(["--binary", b | _]), do: b
  defp parse([_ | rest]), do: parse(rest)
  defp parse([]), do: nil
end
